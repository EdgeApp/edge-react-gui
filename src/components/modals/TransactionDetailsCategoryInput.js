// @flow
import { hook } from 'cavy'
import * as React from 'react'
import { StyleSheet, View } from 'react-native'
import { type AirshipBridge } from 'react-native-airship'

import s from '../../locales/strings.js'
import { FormattedText } from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { TouchableWithoutFeedback } from '../../types/reactNative.js'
import { splitTransactionCategory } from '../../util/utils.js'
import { AirshipModal } from '../common/AirshipModal.js'
import { FormField, MaterialInputOnWhite } from '../common/FormField.js'
import { SubCategorySelect } from '../common/TransactionSubCategorySelect.js'

type CategoriesType = Array<{
  key: string,
  syntax: string
}>

type Props = {
  bridge: AirshipBridge<{ category: string, subCategory: string } | void>,
  categories: Object,
  subCategories: string[],
  category: string,
  subCategory: string,
  setNewSubcategory: (string, string[]) => void,
  generateTestHook: (id: string, ref: any) => void
}

type State = {
  categories: CategoriesType,
  category: string,
  subCategory: string
}

export class TransactionDetailsCategoryInputComponenet extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    const { category, subCategory } = props
    const categories = this.formattedCategories(props.categories)
    this.state = { categories, category, subCategory }
  }

  formattedCategories = (categories: Object): CategoriesType => {
    return Object.keys(categories).map(key => {
      return {
        key: categories[key].key,
        syntax: categories[key].syntax
      }
    })
  }

  onChangeCategory = (category: string) => {
    this.setState({ category })
  }

  onChangeSubCategory = (subCategory: string) => {
    this.setState({ subCategory })
  }

  onSelectSubCategory = (input: string) => {
    const { bridge, subCategories, setNewSubcategory } = this.props
    const splittedFullCategory = splitTransactionCategory(input)
    const { subCategory } = splittedFullCategory
    const category = splittedFullCategory.category.toLowerCase()
    if (!subCategories.find(item => item === input)) {
      setNewSubcategory(input, subCategories)
    }
    bridge.resolve({ category, subCategory })
  }

  render() {
    const { bridge } = this.props
    const { categories, category, subCategory } = this.state
    return (
      <AirshipModal bridge={bridge} onCancel={() => bridge.resolve()}>
        <TouchableWithoutFeedback onPress={() => bridge.resolve()} ref={this.props.generateTestHook('TransactionDetailsCategoryInput.Close')}>
          <View style={styles.airshipContainer}>
            <FormattedText style={styles.airshipHeader}>{s.strings.transaction_details_category_title}</FormattedText>
            <View style={styles.inputCategoryMainContainter}>
              <FormattedText style={styles.inputCategoryListHeader}>{s.strings.tx_detail_picker_title}</FormattedText>
              <View style={styles.inputCategoryRow}>
                {categories.map(item => {
                  const containterStyle = category === item.key ? styles.inputCategoryContainterSelected : styles.inputCategoryContainter
                  return (
                    <TouchableWithoutFeedback onPress={() => this.onChangeCategory(item.key)} key={item.key}>
                      <View style={containterStyle}>
                        <FormattedText style={styles.inputCategoryText}>{item.syntax}</FormattedText>
                      </View>
                    </TouchableWithoutFeedback>
                  )
                })}
              </View>
            </View>
            <View style={styles.inputSubCategoryContainter}>
              <FormField
                {...MaterialInputOnWhite}
                containerStyle={{
                  ...MaterialInputOnWhite.containerStyle,
                  height: THEME.rem(3.44),
                  width: '100%'
                }}
                autoFocus
                returnKeyType="done"
                autoCapitalize="none"
                label={s.strings.transaction_details_choose_a_sub_category}
                fontSize={THEME.rem(0.9)}
                labelFontSize={THEME.rem(0.65)}
                onChangeText={this.onChangeSubCategory}
                onSubmitEditing={() => bridge.resolve({ category, subCategory })}
                value={subCategory}
              />
            </View>
            <SubCategorySelect
              bottomGap={0}
              onPressFxn={this.onSelectSubCategory}
              enteredSubcategory={subCategory}
              subcategoriesList={this.getSortedSubCategories()}
              categories={this.getSortedCategories()}
            />
          </View>
        </TouchableWithoutFeedback>
      </AirshipModal>
    )
  }

  getSortedCategories = (): string[] => {
    const { categories, category } = this.state
    const selectedCategories = categories.filter(item => item.key === category)
    const filteredCategories = categories.filter(item => item.key !== category)
    const sortedCategories = [...selectedCategories, ...filteredCategories]
    return sortedCategories.map(category => category.key)
  }

  getSortedSubCategories = () => {
    const { categories, subCategories } = this.props
    const { category } = this.state

    const selectedSubcategories = subCategories.filter(subCategory => {
      const splittedSubCategory = subCategory.split(':')
      return splittedSubCategory[0].toLowerCase() === categories[category].syntax.toLowerCase()
    })
    const filteredSubcategories = subCategories.filter(subCategory => {
      const splittedSubCategory = subCategory.split(':')
      return splittedSubCategory[0].toLowerCase() !== categories[category].syntax.toLowerCase()
    })
    return [...selectedSubcategories, ...filteredSubcategories]
  }
}

const rawStyles = {
  airshipContainer: {
    flex: 1,
    padding: THEME.rem(0.8)
  },
  airshipHeader: {
    fontSize: THEME.rem(1.2),
    marginBottom: THEME.rem(1),
    alignSelf: 'center'
  },
  inputCategoryListHeader: {
    fontSize: THEME.rem(0.7),
    marginBottom: THEME.rem(0.3),
    color: THEME.COLORS.SECONDARY
  },
  inputCategoryRow: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  inputCategoryContainter: {
    paddingHorizontal: THEME.rem(0.5),
    paddingVertical: THEME.rem(0.2),
    marginRight: THEME.rem(0.6),
    borderWidth: 1,
    borderColor: THEME.COLORS.TRANSACTION_DETAILS_SECONDARY,
    borderRadius: 3
  },
  inputCategoryContainterSelected: {
    paddingHorizontal: THEME.rem(0.5),
    paddingVertical: THEME.rem(0.2),
    marginRight: THEME.rem(0.6),
    borderWidth: 1,
    borderRadius: 3,
    borderColor: THEME.COLORS.TRANSACTION_DETAILS_SECONDARY,
    backgroundColor: THEME.COLORS.TRANSACTION_DETAILS_SECONDARY
  },
  inputCategoryText: {
    color: THEME.COLORS.SECONDARY,
    fontSize: THEME.rem(0.9)
  },
  inputCategoryMainContainter: {
    marginBottom: THEME.rem(0.8)
  },
  inputSubCategoryContainter: {
    marginTop: THEME.rem(0.8)
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)
export const TransactionDetailsCategoryInput = hook(TransactionDetailsCategoryInputComponenet)
