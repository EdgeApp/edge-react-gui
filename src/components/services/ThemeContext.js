// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
