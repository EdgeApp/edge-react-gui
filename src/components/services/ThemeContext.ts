import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
