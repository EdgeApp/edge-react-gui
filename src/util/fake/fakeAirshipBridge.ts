import { AirshipBridge } from 'react-native-airship'

export const fakeAirshipBridge: AirshipBridge<any> = {
  resolve: value => undefined,
  // eslint-disable-next-line node/handle-callback-err
  reject: error => undefined,
  remove: () => undefined,
  on: (name, callback) => () => undefined,
  onResult: callback => undefined
}
