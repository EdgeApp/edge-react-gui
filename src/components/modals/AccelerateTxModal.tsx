import { EdgeCurrencyWallet, EdgeDenomination, EdgeTransaction } from 'edge-core-js'
import React, { PureComponent } from 'react'
import { Text, View } from 'react-native'
import { AirshipBridge } from 'react-native-airship'

import s from '../../locales/strings'
import { Slider } from '../../modules/UI/components/Slider/Slider'
import { getDisplayDenominationFromState, getExchangeDenominationFromState } from '../../selectors/DenominationSelectors'
import { connect } from '../../types/reactRedux'
import { GuiExchangeRates } from '../../types/types'
import { convertTransactionFeeToDisplayFee } from '../../util/utils'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'
import { ModalFooter, ModalMessage, ModalTitle } from '../themed/ModalParts'
import { ThemedModal } from '../themed/ThemedModal'
import { Tile } from '../tiles/Tile'

interface OwnProps {
  acceleratedTx: EdgeTransaction
  bridge: AirshipBridge<EdgeTransaction | null>
  replacedTx: EdgeTransaction
  wallet: EdgeCurrencyWallet
}
interface StateProps {
  exchangeRates: GuiExchangeRates
}
interface DispatchProps {
  getDisplayDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
  getExchangeDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
}
type Props = OwnProps & StateProps & ThemeProps & DispatchProps

interface State {
  error?: Error
  status: 'confirming' | 'sending' | 'sent'
  mounted: boolean
}

export class AccelerateTxModalComponent extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      error: undefined,
      status: 'confirming',
      mounted: true
    }
  }

  signBroadcastAndSave = async () => {
    const { acceleratedTx, wallet } = this.props

    this.setState({ status: 'sending' })

    try {
      // Sign, broadcast, and save the accelerated transaction
      const signedTx = await wallet.signTx(acceleratedTx)
      await wallet.broadcastTx(signedTx)
      await wallet.saveTx(signedTx)

      if (this.state.mounted) {
        this.setState({ status: 'sent' })
      }
      this.closeModal(signedTx)
    } catch (error: any) {
      if (this.state.mounted) {
        this.closeModal(null, error)
      }
    }
  }

  closeModal = (signedTx: EdgeTransaction | null, err?: Error) => {
    this.setState({
      mounted: false
    })
    if (err != null) this.props.bridge.reject(err)
    else this.props.bridge.resolve(signedTx)
  }

  handleCancel = () => {
    this.closeModal(null)
  }

  handleConfirmation = async () => {
    await this.signBroadcastAndSave()
  }

  getTxFeeDisplay = (edgeTransaction: EdgeTransaction): string => {
    const { exchangeRates, wallet, getDisplayDenomination, getExchangeDenomination } = this.props

    const feeDisplayDenomination = getDisplayDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const feeDefaultDenomination = getExchangeDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const transactionFee = convertTransactionFeeToDisplayFee(wallet, exchangeRates, edgeTransaction, feeDisplayDenomination, feeDefaultDenomination)

    const feeSyntax = `${transactionFee.cryptoSymbol ?? ''} ${transactionFee.cryptoAmount} (${transactionFee.fiatSymbol ?? ''} ${transactionFee.fiatAmount})`

    return feeSyntax
  }

  render() {
    const { acceleratedTx, bridge, replacedTx, theme } = this.props
    const { error, status } = this.state

    const styles = getStyles(theme)

    const oldFee = this.getTxFeeDisplay(replacedTx)
    const newFee = this.getTxFeeDisplay(acceleratedTx)

    const isSending = status === 'sending'

    return (
      <ThemedModal bridge={bridge} onCancel={this.handleCancel}>
        <ModalTitle>{s.strings.transaction_details_accelerate_transaction_header}</ModalTitle>
        <ModalMessage>{s.strings.transaction_details_accelerate_transaction_instructional}</ModalMessage>
        <View style={styles.container}>
          <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_old_fee_title} body={oldFee} />
          {newFee == null ? null : <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_new_fee_title} body={newFee} />}
        </View>
        {error == null ? null : (
          <View style={styles.error}>
            <Text style={styles.errorText} numberOfLines={3}>
              {error.message}
            </Text>
          </View>
        )}
        <View style={styles.container}>
          <Slider
            disabled={isSending || !!error}
            onSlidingComplete={this.handleConfirmation}
            showSpinner={isSending}
            disabledText={s.strings.transaction_details_accelerate_transaction_slider_disabled}
          />
        </View>
        <ModalFooter onPress={this.handleCancel} />
      </ThemedModal>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  loadingContianer: {
    height: theme.rem(20)
  },
  error: {
    marginVertical: theme.rem(0.5)
  },
  errorText: {
    textAlign: 'center',
    color: theme.negativeText
  },
  loading: {
    flex: 1,
    alignSelf: 'center'
  },
  container: {
    maxHeight: theme.rem(20),
    marginVertical: theme.rem(0.5)
  }
}))

export const AccelerateTxModal = connect<StateProps, DispatchProps, OwnProps>(
  (state, ownProps) => ({
    exchangeRates: state.exchangeRates
  }),
  dispatch => ({
    getDisplayDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
    },
    getExchangeDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getExchangeDenominationFromState(pluginId, currencyCode))
    }
  })
)(withTheme(AccelerateTxModalComponent))
