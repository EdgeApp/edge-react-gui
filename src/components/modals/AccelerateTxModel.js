// @flow

import type { EdgeCurrencyWallet, EdgeDenomination, EdgeSpendInfo, EdgeTransaction } from 'edge-core-js'
import React, { PureComponent } from 'react'
import { ActivityIndicator, Text, View } from 'react-native'
import { type AirshipBridge } from 'react-native-airship'

import { playSendSound } from '../../actions/SoundActions.js'
import { TRANSACTION_DETAILS } from '../../constants/SceneKeys.js'
import s from '../../locales/strings.js'
import { Slider } from '../../modules/UI/components/Slider/Slider.js'
import { getDisplayDenominationFromState, getExchangeDenominationFromState } from '../../selectors/DenominationSelectors.js'
import { type TestProps, connect } from '../../types/reactRedux.js'
import { Actions } from '../../types/routerTypes.js'
import { type GuiExchangeRates } from '../../types/types.js'
import { convertTransactionFeeToDisplayFee } from '../../util/utils.js'
import { showError, showToast, showWarning } from '../services/AirshipInstance.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { ModalCloseArrow, ModalMessage, ModalTitle } from '../themed/ModalParts.js'
import { ThemedModal } from '../themed/ThemedModal.js'
import { Tile } from '../tiles/Tile'

type Status = 'confirming' | 'sending' | 'sent'

type OwnProps = {
  bridge: AirshipBridge<Status>,
  edgeTransaction: EdgeTransaction,
  wallet: EdgeCurrencyWallet
}
type StateProps = {
  exchangeRates: GuiExchangeRates
}
type DispatchProps = {
  getDisplayDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination,
  getExchangeDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
}
type Props = OwnProps & StateProps & ThemeProps & DispatchProps & TestProps

type State = {
  edgeUnsignedTransaction?: EdgeTransaction,
  error?: Error,
  status: Status,
  mounted: boolean
}

export class AccelerateTxModelComponent extends PureComponent<Props, State> {
  constructor() {
    super()

    this.state = {
      edgeUnsignedTransaction: undefined,
      error: undefined,
      status: 'confirming',
      mounted: true
    }
  }

  componentDidMount() {
    this.makeRbfTransaction()
  }

  makeRbfTransaction = async () => {
    const { edgeTransaction, wallet } = this.props

    const edgeTransactionSpendTargets = edgeTransaction.spendTargets

    if (edgeTransactionSpendTargets && edgeTransactionSpendTargets.length) {
      // Currency code for the new tx is the same as the first spend target
      const { currencyCode } = edgeTransactionSpendTargets[0]
      // Map the EdgeTransaction.spendTargets type to EdgeSpendTargets
      const spendTargets = edgeTransactionSpendTargets.map(spendTarget => ({
        nativeAmount: spendTarget.nativeAmount,
        publicAddress: spendTarget.publicAddress,
        uniqueIdentifier: spendTarget.uniqueIdentifier
      }))
      const {
        // Use the txid from the replaced transaction as the rbfTxid
        txid: rbfTxid,
        // Copy replaced transaction's metadata and swapInfo to the RBF transaction
        metadata,
        swapData
      } = edgeTransaction

      const edgeSpendInfo: EdgeSpendInfo = {
        currencyCode,
        spendTargets,
        rbfTxid,
        metadata,
        swapData
      }

      try {
        const edgeUnsignedTransaction = await wallet.makeSpend(edgeSpendInfo)

        this.setState({
          edgeUnsignedTransaction
        })
      } catch (error) {
        this.setState({
          error
        })
      }
    } else {
      const error = new Error('Missing spend target data.')
      this.setState({
        error
      })
    }
  }

  signBroadcastAndSaveRbf = async () => {
    const { wallet } = this.props
    const { edgeUnsignedTransaction } = this.state

    if (edgeUnsignedTransaction) {
      let edgeSignedTransaction: EdgeTransaction = edgeUnsignedTransaction

      this.setState({ status: 'sending' })

      try {
        // Sign, broadcast, and save the RBF transaction
        edgeSignedTransaction = await wallet.signTx(edgeUnsignedTransaction)
        edgeSignedTransaction = await wallet.broadcastTx(edgeSignedTransaction)
        await wallet.saveTx(edgeSignedTransaction)

        if (this.state.mounted) {
          playSendSound().catch(error => console.log(error)) // Fail quietly

          this.setState({ status: 'sent' })

          showToast(s.strings.transaction_success_message)

          Actions.replace(TRANSACTION_DETAILS, { edgeTransaction: edgeSignedTransaction })
        } else {
          showWarning(s.strings.transaction_success_message)
        }
      } catch (error) {
        console.log(error)

        if (this.state.mounted) {
          this.setState({ status: 'confirming' })
          showError(error)
        }
      }
    } else {
      throw new Error(s.strings.invalid_spend_request)
    }
  }

  closeModal = () => {
    this.setState({
      mounted: false
    })
    this.props.bridge.resolve(this.state.status)
  }

  handleConfirmation = async () => {
    await this.signBroadcastAndSaveRbf()
    this.closeModal()
  }

  getTxFeeDisplay = (edgeTransaction: EdgeTransaction): string => {
    const { exchangeRates, wallet, getDisplayDenomination, getExchangeDenomination } = this.props

    const feeDisplayDenomination = getDisplayDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const feeDefaultDenomination = getExchangeDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const transactionFee = convertTransactionFeeToDisplayFee(wallet, exchangeRates, edgeTransaction, feeDisplayDenomination, feeDefaultDenomination)

    const feeSyntax = `${transactionFee.cryptoSymbol ?? ''} ${transactionFee.cryptoAmount} (${transactionFee.fiatSymbol ?? ''} ${transactionFee.fiatAmount})`

    return feeSyntax
  }

  render() {
    const { bridge, edgeTransaction, theme } = this.props
    const { error, status, edgeUnsignedTransaction } = this.state

    const styles = getStyles(theme)

    const oldFee = this.getTxFeeDisplay(edgeTransaction)
    const newFee = edgeUnsignedTransaction != null ? this.getTxFeeDisplay(edgeUnsignedTransaction) : ''

    const isSending = status === 'sending'

    return (
      <ThemedModal bridge={bridge} onCancel={this.closeModal}>
        {edgeUnsignedTransaction || error ? (
          <>
            <ModalTitle>{s.strings.transaction_details_accelerate_transaction_header}</ModalTitle>
            <ModalMessage>{s.strings.transaction_details_accelerate_transaction_instructional}</ModalMessage>
            <View style={styles.container}>
              <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_old_fee_title} body={oldFee} />
              {!!newFee && <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_new_fee_title} body={newFee} />}
            </View>
            {error && (
              <View style={styles.error}>
                <Text style={styles.errorText} numberOfLines={3}>
                  {error.message}
                </Text>
              </View>
            )}
            <View style={styles.container}>
              <Slider
                disabled={isSending || !!error}
                onSlidingComplete={this.handleConfirmation}
                showSpinner={isSending}
                disabledText={s.strings.transaction_details_accelerate_transaction_slider_disabled}
              />
            </View>
            <ModalCloseArrow onPress={this.closeModal} ref={this.props.generateTestHook('AccelerateTxModal.Close')} />
          </>
        ) : (
          <View style={styles.loadingContianer}>
            <ActivityIndicator color={theme.primaryText} style={styles.loading} size="large" />
          </View>
        )}
      </ThemedModal>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  loadingContianer: {
    height: theme.rem(20)
  },
  error: {
    marginVertical: theme.rem(0.5)
  },
  errorText: {
    textAlign: 'center',
    color: theme.negativeText
  },
  loading: {
    flex: 1,
    alignSelf: 'center'
  },
  container: {
    maxHeight: theme.rem(20),
    marginVertical: theme.rem(0.5)
  }
}))

export const AccelerateTxModel = connect<StateProps, DispatchProps, OwnProps & TestProps>(
  (state, ownProps) => ({
    exchangeRates: state.exchangeRates
  }),
  dispatch => ({
    getDisplayDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
    },
    getExchangeDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getExchangeDenominationFromState(pluginId, currencyCode))
    }
  })
)(withTheme(AccelerateTxModelComponent))
