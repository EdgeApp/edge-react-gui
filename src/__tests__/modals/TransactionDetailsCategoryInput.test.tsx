/* globals describe it expect */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { TransactionDetailsCategoryInput } from '../../components/modals/TransactionDetailsCategoryInput'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeAirshipBridge } from '../../util/fake/fakeAirshipBridge'

describe('TransactionDetailsCategoryInput', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      bridge: fakeAirshipBridge,
      categories: {
        exchange: {
          syntax: 'Hello exchange',
          key: 'exchange'
        },
        expense: {
          syntax: 'Hello expense',
          key: 'expense'
        },
        transfer: {
          syntax: 'Hello transfer',
          key: 'transfer'
        },
        income: {
          syntax: 'Hello income',
          key: 'income'
        }
      },
      subCategories: ['exchange', 'income', 'expense'],
      category: 'exchange',
      subCategory: '',
      // eslint-disable-next-line no-empty-pattern
      setNewSubcategory: (string, []) => undefined,
      theme: getTheme()
    }
    const actual = renderer.render(<TransactionDetailsCategoryInput {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
