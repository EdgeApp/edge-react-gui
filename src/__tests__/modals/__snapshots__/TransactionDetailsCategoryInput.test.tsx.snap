// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetailsCategoryInput should render with loading props 1`] = `
<AirshipModal
  bridge={
    Object {
      "on": [Function],
      "onResult": [Function],
      "reject": [Function],
      "remove": [Function],
      "resolve": [Function],
    }
  }
  onCancel={[Function]}
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "flex": 1,
          "padding": 18,
        }
      }
    >
      <FormattedText
        style={
          Object {
            "alignSelf": "center",
            "fontSize": 27,
            "marginBottom": 22,
          }
        }
      >
        Category
      </FormattedText>
      <View
        style={
          Object {
            "marginBottom": 18,
          }
        }
      >
        <FormattedText
          style={
            Object {
              "color": "#0E4B75",
              "fontSize": 16,
              "marginBottom": 7,
            }
          }
        >
          Choose a Category:
        </FormattedText>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "backgroundColor": "#A4C7DF",
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Hello exchange
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Hello expense
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Hello transfer
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Hello income
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </View>
      <View
        style={
          Object {
            "marginTop": 18,
          }
        }
      >
        <FormField
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={true}
          baseColor="#0D2145"
          containerStyle={
            Object {
              "height": 77,
              "position": "relative",
              "width": "100%",
            }
          }
          errorColor="#E85466"
          fontSize={20}
          keyboardType="default"
          label="Choose a sub-category"
          labelFontSize={15}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          returnKeyType="done"
          textColor="#25292C"
          tintColor="#0E4B75"
          titleTextStyle={Object {}}
          value=""
        />
      </View>
      <SubCategorySelect
        bottomGap={0}
        categories={
          Array [
            "exchange",
            "expense",
            "transfer",
            "income",
          ]
        }
        enteredSubcategory=""
        onPressFxn={[Function]}
        subcategoriesList={
          Array [
            "exchange",
            "income",
            "expense",
          ]
        }
      />
    </View>
  </TouchableWithoutFeedback>
</AirshipModal>
`;
